numpy/core/src/umath/ufunc_object.c

```diff
@@ -1738,6 +1753,156 @@ make_arr_prep_args(npy_intp nin, PyObject *args, PyObject *kwds)
+ * axes and then the core axes the gufunc operates on.
+ *
+ * Returns 0 on success, and -1 on failure
+ */
+static int
+_parse_axes_arg(PyUFuncObject *ufunc, PyObject *axes, PyArrayObject **op,
+                int broadcast_ndim, int **remap_axis) {
+    int nin = ufunc->nin;
+    int nout = ufunc->nout;
+    int nop = nin + nout;
+    int iop, list_size;
+    PyObject *op_axes_tuple;
+
+    if (!PyList_Check(axes)) {
+        PyErr_SetString(PyExc_TypeError, "axes should be a list.");
+        return -1;
+    }
+    list_size = PyList_Size(axes);
+    if (list_size != nop) {
+        if (list_size != nin || _has_output_coredims(ufunc)) {
+            PyErr_Format(PyExc_ValueError,
+                         "axes should be a list with an entry for all "
+                         "%d inputs and outputs; entries for outputs can only "
+                         "be omitted if none of them has core axes.",
+                         nop);
+            return -1;
+        }
+        for (iop = nin; iop < nop; iop++) {
+            remap_axis[iop] = NULL;
+        }
+    }
+    for (iop = 0; iop < list_size; ++iop) {
+        int op_ndim, op_ncore, op_nbroadcast;
+        int have_seen_axis[NPY_MAXDIMS] = {0};
+        PyObject *axis_item;
+        int axis, op_axis;
+
+        op_ncore = ufunc->core_num_dims[iop];
+        if (op[iop] != NULL) {
+            op_ndim = PyArray_NDIM(op[iop]);
+            op_nbroadcast = op_ndim - op_ncore;
+        }
+        else {
+            op_nbroadcast = broadcast_ndim;
+            op_ndim = broadcast_ndim + op_ncore;
+        }
+        op_axes_tuple = PyList_GetItem(axes, iop);
+        if (op_axes_tuple == NULL) {
+            return -1;
+        }
+        /* If not a tuple already, make it one; content is checked later */
+        if (!PyTuple_Check(op_axes_tuple)) {
+            op_axes_tuple = PyTuple_Pack(1, op_axes_tuple);
+            if (op_axes_tuple == NULL) {
+                return -1;
+            }
+        }
+        else {
+            Py_INCREF(op_axes_tuple);
+        }
+        if (PyTuple_Size(op_axes_tuple) != op_ncore) {
+            if (op_ncore == 1) {
+                PyErr_Format(PyExc_ValueError,
+                             "axes item %d should be a tuple with a "
+                             "single element, or an integer", iop);
+            }
+            else {
+                PyErr_Format(PyExc_ValueError,
+                             "axes item %d should be a tuple with %d "
+                             "elements", iop, op_ncore);
+            }
+            goto fail;
+        }
+        /*
+         * Now create the remap, starting with the core dimensions, and then
+         * adding the remaining broadcast axes that are to be iterated over.
+         */
+        for (axis = op_nbroadcast; axis < op_ndim; axis++) {
+            axis_item = PyTuple_GetItem(op_axes_tuple, axis - op_nbroadcast);
+            if (axis_item == NULL) {
+                goto fail;
```

Comment:
My previous comment was wrong, if the index is correct this can never fail.

Reply:
OK, removed those checks again.